### ISSUE TITLE
Rainbow didn't issue a battery charge plan for battery set perturbation

### ISSUE THREAD

**Issue by pooyanjamshidi** [cmu-mars/issues#85]
The rainbow didn't issue a plan for recharge.

**Description**: This was a C case test with the mission specified in [test-ready-c1c.json](https://github.mit.edu/brass/cmu-robotics/blob/development/cp1/ta/missions/test-ready-c1c.json)
**Issue**: I expected to see a plan for going to charging station after we set the battery low
**Start conditions**: This is the 3a case where we set the battery to a low value but good enough for reaching the charging station. 
**Steps to reproduce**: 
```bash
curl -X POST "http://0.0.0.0:8080/start" -H "accept: application/json
curl -X POST "http://0.0.0.0:8080/perturb/battery" -H "accept: application/json" -H "Content-Type: application/json" -d "{ \"charge\": 3121.73}"
```

**Observations**: I didn't see any adaptation for battery charge.
[rainbow_c3a1.log](https://github.com/cmu-mars/issues/files/2002279/rainbow_c3a1.log)
[commnad_line_c3a.log](https://github.com/cmu-mars/issues/files/2002280/commnad_line_c3a.log)

**Comment by pooyanjamshidi** [comment]
@schmerl would you please have a look at the log? maybe I could not find the plan in the log, but as far as I saw in the command line there was no plan for going to charge station.

**Comment by pooyanjamshidi** [comment]
I also tested with higher battery perturbations and observed the same: no adaptation.
```
curl -X POST "http://0.0.0.0:8080/start" -H "accept: application/json
curl -X POST "http://0.0.0.0:8080/perturb/battery" -H "accept: application/json" -H "Content-Type: application/json" -d "{ \"charge\": 7121.73}"
```

**Comment by javier-camara** [comment]
I am trying to look into this, but I am missing some information. In the test you refer to, there are several locations that the robot has to visit. Which one is failing?
Concretely, I would need to know start and end location for the specific task (not the overall mission) where this is failing, as well as the level of battery that the robot has at the start of the task.

**Comment by schmerl** [comment]
You couldn't find the plan. There are two interesting things in the log that I can see:

```
INFO  ESEBMasterReportingPort - ANALYSIS: BRASS Accuracy Evaluator: Do not have enough battery.
INFO  ESEBMasterReportingPort - 
INFO  ESEBMasterReportingPort - ANALYSIS: BRASS Accuracy Evaluator: Inserting node 'l13' at (22.7818915973, 10.7212516309) between l1 and l3
```
Indicates that the battery charge was deemed to be too low to complete the mission.

```
INFO  ESEBMasterReportingPort - MODEL: Executing O[RainbowState:RainbowState/setModelProblem.([LOW_ON_BATTERY])]
```

Indicates that the problem is recorded properly, meaning that the planner should now know there is a problem.
There's a whole bunch of output from prism, culminating in it producing the plan:

```
INFO  ESEBMasterReportingPort - EFFECTOR_MANAGER: Attempting E[changeInstructionGraph@172.18.0.2] ([P(V(1, do SetCP1Config(7) then 2),
V(2, do MoveAbsH(22.97, 10.87, 0.63, 0.0164) then 3)::
V(3, end)::
nil)])
```

@javier-camara Can you check the prism output to see if it makes sense? To me the power numbers look low enough (e.g. 286) that I would be surprised if it was able to get to a charging station.

**Comment by pooyanjamshidi** [comment]
@javier-camara here I also included the command line logs:
[rainbow_c3a1.log](https://github.com/cmu-mars/issues/files/2002256/rainbow_c3a1.log)
[commnad_line_c3a.log](https://github.com/cmu-mars/issues/files/2002257/commnad_line_c3a.log)

So as you can see, in this case, it finished the `l4->l2` task and it is heading to `l5`, but this fails, also the battery is around `1200mwh` at the start of `l2->l5`. Note that this is pretty low and it was also low when it started `l4->l2` (`3300mwh`).

**Comment by pooyanjamshidi** [comment]
Also for your information, the battery capacity is `32560.0mwh` and the perturbation happened right at the beginning, at `"sim-time": 248`.

**Comment by ivoysey** [comment]
sim-time 248 isn't the beginning -- that's 248 simulated seconds into the
experiment.

**Comment by pooyanjamshidi** [comment]
it is just 1 second after the first `curl` that started the mission.

**Comment by ivoysey** [comment]
That's not relevant; there's no real relationship between wall time and sim time, and sim time is the one that matters. This maybe related to https://github.com/cmu-mars/issues/issues/71

**Comment by pooyanjamshidi** [comment]
@ivoysey I don't get how your comment is related to this issue? can you clarify how does it help? I just provided @javier-camara information that he may/may not need.

**Comment by schmerl** [comment]
Time is a real issue here (by which I mean for us and BRASS in general). 248 is *not* 248 (simulated) seconds (ss) after the robot started its mission, it's 248ss after Gazebo started. So, if I understand correctly, you did "start" and then "straight afterward" you did "perturb". So start might have happened at 246ss and perturb at 248ss.

**Comment by javier-camara** [comment]
Independently of the timing issues, this situation has lead me to discover a bug in the planning model generation. There is a corner case that wasn't being considered: the robot would arrive to a charging station with very little battery, and, in the way in which the guards are structured, the recharging tactic won't happen under those conditions. I'm currently working on a fix.

**Comment by pooyanjamshidi** [comment]
@schmerl time is important not only in BRASS but for the human being, for the universe, for etc, but how time is relevant for solving this very specific issue is a mystery for me, probably my fault giving info that was not necessary ;)

**Comment by javier-camara** [comment]
I have committed some code that should fix this issue, but let's wait and close the issue once we have tested the fix.

**Comment by pooyanjamshidi** [comment]
@schmerl I would like to test this, would you please fix the issue in #89 to allow me to run this charging plan test?

**Comment by schmerl** [comment]
@pooyanjamshidi  wrote in #89:

@javier-camara I see NullPointerException in EnergyConsumptionPredictor in the command-line log, and the planner apparently didn't come with a plan to go to charging, here are the logs. Can you look into this? I'm in my office if you want to go throughout together.

[rainbow_c2a4.log](https://github.com/cmu-mars/issues/files/2006379/rainbow_c2a4.log)
[commnad_line_c2a4.log](https://github.com/cmu-mars/issues/files/2006380/commnad_line_c2a4.log)

**Comment by schmerl** [comment]
@pooyanjamshidi  Is this the right issue? Isn't this to do with the battery perturbation issue?

In any case, what makes you say that the planner doesn't come up with a plan that includes charging? I see:

```
(V(1, do SetCP1Config(1) then 2),
V(2, do MoveAbsH(-21.08, 0.00, 0.12, -1.5522) then 3)::
V(3, do MoveAbsH(-20.86, -11.62, 0.12, -0.0123) then 4)::
V(4, do MoveAbsH(1.03, -11.89, 0.12, 0.0000) then 5)::
V(5, do Charge(15.00) then 6)::
V(6, do MoveAbsH(23.19, -11.89, 0.12, 0.0000) then 7)::
V(7, end)::
nil)])
```

There are two things going on here, and I think you quit prematurely:

1. Does the IG get executed? It is outside of Rainbow and so is not affected by exceptions in Rainbow.
2. Rainbow is throwing an exception in the energy consumption part, which is not part of the planner.

**Comment by schmerl** [comment]
@pooyanjamshidi @javier-camara Let's the three of us try to meet tomorrow. There are some assumptions in the IGWaypointAnalyzer that may be invalidated because of the time taken to come up with a plan.

**Comment by pooyanjamshidi** [comment]
@schmerl right, I missed the charge() when I searched through the log.

**Comment by pooyanjamshidi** [comment]
@schmerl In the last log, we do not have any exception regarding `IGWaypointAnalyzer`, however, the exception is related to `EnergyConsumptionPredictor`. We will discuss both in the meeting today, but wondering can you please fix the later then I can run few tests before meeting?

**Comment by schmerl** [comment]
The latter is related to the former.

**Comment by schmerl** [comment]
OK, I've checked in a build that tries to do some more tracing and maybe will fix the problem. @pooyanjamshidi Can you give it a try before we meet and attach the logs? Both rainbow and console. Rebuild cp1_rb.

**Comment by pooyanjamshidi** [comment]
@schmerl :
[rainbow_c2a5.log](https://github.com/cmu-mars/issues/files/2009697/rainbow_c2a5.log)
[commnad_line_c2a5.log](https://github.com/cmu-mars/issues/files/2009698/commnad_line_c2a5.log)

**Comment by schmerl** [comment]
So this is the tracing that I am getting:

```
INFO  Analyzer - IG to Waypoint: Robot is not at a location in the map: (-18.4509321135, 11.0690397984)
INFO  ESEBMasterReportingPort - ANALYSIS: IG to Waypoint: Robot is not at a location in the map: (-18.4509321135, 11.0690397984)
INFO  ESEBMasterReportingPort - 
INFO  Analyzer - IG to Waypoint: Robot is not a a location in the map
INFO  ESEBMasterReportingPort - ANALYSIS: IG to Waypoint: Robot is not a a location in the map
INFO  ESEBMasterReportingPort - 
INFO  Analyzer - IG to Waypoint: The source node is not set, using current pose to guess where robot is coming from
INFO  ESEBMasterReportingPort - ANALYSIS: IG to Waypoint: The source node is not set, using current pose to guess where robot is coming from
INFO  ESEBMasterReportingPort - 
INFO  Analyzer - IG to Waypoint: Guessing that current source is l2
INFO  ESEBMasterReportingPort - ANALYSIS: IG to Waypoint: Guessing that current source is l2
INFO  ESEBMasterReportingPort - 
INFO  Analyzer - IG to Waypoint: IG processor encountered an error
INFO  ESEBMasterReportingPort - ANALYSIS: IG to Waypoint: IG processor encountered an error
INFO  ESEBMasterReportingPort - 
```
I suspect that the problem is that by the time the planner comes up with a plan, the robot has moved away from where it was when the plan started. This violates some assumptions within the models of rainbow (@bisc !) and so some exceptions are thrown. 

One workaround is to make the energy consumer robust to this. Another is what I have done. Another one is to stop the robot before we start planning.

**Comment by schmerl** [comment]
@pooyanjamshidi : Some more robustness for the energy predictor and less waiting in the IG analyzer has been implemented. Rebuild cp1_rb.

**Comment by pooyanjamshidi** [comment]
I reran this again, but the planner didn't come up with charging instructions, what's wrong here @javier-camara and @schmerl ? note the same setting as above.

[rainbow_c2a9.log](https://github.com/cmu-mars/issues/files/2011093/rainbow_c2a9.log)
[commnad_line_c2a9.log](https://github.com/cmu-mars/issues/files/2011094/commnad_line_c2a9.log)

**Comment by schmerl** [comment]
Maybe config 7 is efficient enough?

**Comment by pooyanjamshidi** [comment]
@schmerl The robot ran out of battery and it failed the mission!

**Comment by schmerl** [comment]
Do you log when that happens?

**Comment by pooyanjamshidi** [comment]
Exactly after the last log line in the command line log, after that the robot shuts down and some exceptions appear.

**Comment by pooyanjamshidi** [comment]
@schmerl I push the latest to docker hub so you can have a look.

**Comment by schmerl** [comment]
@javier-camara this does look a bit strange given the initial conditions
and the plan selected:


Model constants:
INITIAL_CONFIGURATION=-1,INITIAL_LOCATION=1,TARGET_LOCATION=4,INITIAL_BATTERY=735,INITIAL_HEADING=2
Building reward structure...

Starting expected reachability (max)...
Starting Prob1 (min)...
Prob1 (min) took 16 iterations and 0.004 seconds.
target=14, inf=0, rest=184
Starting value iteration (max)...
Value iteration (max) took 1 iterations and 0.0 seconds.
Expected reachability took 0.006 seconds.

Value in the initial state: 0.0

Time for model checking: 0.008 seconds.

Result: 0.0 (value in the initial state)

Exporting strategy as actions to file
"/home/mars/das/rainbow-brass/targets/brass-p2-cp1/prism/0.act"...

Exporting strategy as an induced model to file
"/home/mars/das/rainbow-brass/targets/brass-p2-cp1/prism/0.ind"...
Selected candidate policy:
/home/mars/das/rainbow-brass/targets/brass-p2-cp1/prism//0
Score: 0.0 Energy: 0.0 Time: 12.0
INFO  ESEBMasterReportingPort - ADAPTATION_MANAGER: Planner chooses
the plan [t_set_sol_7, l2_to_l10, l10_to_l7, l7_to_l6, l6_to_l5]

**Comment by javier-camara** [comment]
@pooyanjamshidi , what is the battery discharge rate that appears in the JSON for sol_7?

**Comment by pooyanjamshidi** [comment]
@javier-camara I don't know because from run to run is different, I can print it out after the learning is done, so in every run, we can have the list of configurations.

**Comment by javier-camara** [comment]
It would be useful to find out what's going on. The plan generated depends very much on this information. Could you try to replicate the test and provide the list of configs with their discharge rates? (Or at least, for the one selected by the planner)

**Comment by pooyanjamshidi** [comment]
@javier-camara configuration data in the command-line log at the beginning:
[rainbow_c2a10.log](https://github.com/cmu-mars/issues/files/2011302/rainbow_c2a10.log)
[commnad_line_c2a10.log](https://github.com/cmu-mars/issues/files/2011303/commnad_line_c2a10.log)

**Comment by pooyanjamshidi** [comment]
@javier-camara it doesn't seem a right adaptation to me given the power load, also as you see it failed and robot ran out of battery.

**Comment by pooyanjamshidi** [comment]
@javier-camara the reconfiguration decrease the power load, but what I expected to see is to go to charging in the previous task, i.e., `l4->l2`, because when it starts the task its battery level is `~3600` which is pretty low, although it is enough to get to `l2`, it is not enough for any adaptation. Do you think the same?

**Comment by javier-camara** [comment]
There may be more things, but I have just realized that the log has two different fields, which are power_load and power_load_w (which is in mwhr, I assume). The code assumes a JSON that reads power_load, not power_load_w (the format change in the JSON is something of which I never received any notification).

The format on which we agreed is: 
{
  "configurations": [
    {
      "config_id": "12", 
      "power_load": 60, 
      "speed": 0.74
    }, 
...
}

Where power_load is in mwhr, and speed in m/s.

**Comment by javier-camara** [comment]
If my assumption is right, I can change the code to read the value in power_load_w instead. If you confirm, I will make the change.

**Comment by pooyanjamshidi** [comment]
@javier-camara `power_load` is indeed in `mwhr`, I use `power_load_w` for other purposes which is not relevant for the planner, etc. So, don't worry about that.

**Comment by javier-camara** [comment]
But if power_load is in mwhr, the values are too low for recharging to be required, right? Configuration 7 has a power_load of 8.77 mwhr, and a speed of 0.74 m/s. That means that the energy consumed per meter is approx 11 mwhr, and assuming an initial battery charge like the one sent above by Bradley (735 mwhr, which btw, is very low, compared to the overall the battery capacity of 32.5 whr), the robot can still travel about 62 meters without recharging. That is a super-efficient configuration, compared to what we are used to see. From the planner perspective, that would make sense. Is the battery simulator consistent with these numbers, or are there any mistakes in my calculations?

**Comment by pooyanjamshidi** [comment]
@javier-camara It appears that the reconfiguration is not enacted, see at the end of the log, we see `The current configuration was requested: 0` this means still operates on the default configuration. @schmerl is this an IG bug? if the reconfiguration is actually called by rainbow, then IG didn't execute it, otherwise, we should have seen `Setting configuration to `.

**Comment by pooyanjamshidi** [comment]
@javier-camara Also, why in every single run it wants to reconfigure to configuration 7? while there are some other configurations with a lower load?

**Comment by schmerl** [comment]
@pooyanjamshidi This line says that the last instruction received by the IG was the one given by Rainbow:

```
[INFO] [1526519957.697073, 410.455000]: BRASS | IG | Recieved a new goal: P(V(1, do SetCP1Config(7) then 2),V(2, do MoveAbsH(-21.08, 11.08, 0.75, -1.5708) then 3)::V(3, do MoveAbsH(-21.08, 0.00, 0.75, -1.5476) then 4)::V(4, do MoveAbsH(-20.81, -11.62, 0.75, -0.0123) then 5)::V(5, do MoveAbsH(1.08, -11.89, 0.75, 0.0000) then 6)::V(6, do MoveAbsH(23.24, -11.89, 0.75, 0.0000) then 7)::V(7, end)::nil)
```

These lines say that the setCP1Config seems to have been executed:
```
cp1_ta    | [INFO] [1526519957.699916, 410.455000]: Executing the graph
cp1_ta    | Executing 8
cp1_ta    | [INFO] [1526519957.701176, 410.455000]: 7.0
```
But the configuration still seems to be reported as 0. Problem with the SetCP1Config in `cp1_instructions.py`?

**Comment by javier-camara** [comment]
@pooyanjamshidi the power_load is not the only parameter that determines the energy consumption. You have to factor in speed to determine energy consumed per meter, so finding out which configuration is more efficient just by looking at power_load is not obvious.

**Comment by schmerl** [comment]
@pooyanjamshidi said:
```
 but what I expected to see is to go to charging in the previous task, i.e., l4->l2, because when it starts the task its battery level is ~3600 which is pretty low, although it is enough to get to l2, it is not enough for any adaptation. Do you think the same?
```
But, this is not a correct expectation, right? The planner is only considering the current task (got to l2) and if it can get there without charging (because of timeliness) it will. This means that it may not have enough power to get to a charging station in the next task.

**Comment by javier-camara** [comment]
@schmerl, you are right. Part of this is that the weights for the timeliness and energy consumption are fixed to 0.5 for the time being (in rainbow/deployments/rainbow-brass/src/main/java/org/sa/rainbow/brass/plan/p2_cp1/DecisionEngineCP1.java), so the robot does not necessarily favor energy efficiency. If we modify the weights to be e.g. 0.9 for energy efficiency, vs. 0.1 for timeliness (as we did in the experiment on my machine the other day @pooyanjamshidi ), I would expect to see a recharging tactic showing up in the plan.

**Comment by javier-camara** [comment]
Just to clarify, by "energy efficiency" I mean maximization of the remaining battery energy by the end of the task.

**Comment by pooyanjamshidi** [comment]
@schmerl last night I doubled check `cp1_instructions.py` and it seems alright, so you are 100% sure the method `status, msg = self.cp1.set_config(config)` got executed?

**Comment by pooyanjamshidi** [comment]
@javier-camara Would you please make the utilities parameterized so that I can play with it to find a balance that it succeed a mission now for RR2? see #94 or if this is a too much change now, probably change the weight to 0.9-0.1 for now?

**Comment by pooyanjamshidi** [comment]
@javier-camara since you are leaving today, probably simply change that to 0.9 - 0.1.

**Comment by javier-camara** [comment]
@pooyanjamshidi changed priorities to 0.9 -0.1 in the repo.

**Comment by pooyanjamshidi** [comment]
@schmerl can you please build rb given javier changes?

**Comment by schmerl** [comment]
@pooyan. Not yet. Let's deal with #84 first. Did you check the setting of the config mentioned a few comments ago?

**Comment by pooyanjamshidi** [comment]
@schmerl yeah, I mentioned that I checked `set_config` and it seems fine, but I believe `ig_server` doesn't call that method, could you check? because I put a comment in the first line of the method and I don't see it int he log...

**Comment by schmerl** [comment]
@pooyanjamshidi Checked in a change to ig_server that should have the trace messages appear. And it should be being called.

**Comment by pooyanjamshidi** [comment]
@javier-camara and @schmerl logs for battery perturbations, the exception is strange:
[commnad_line_c2a14.log](https://github.com/cmu-mars/issues/files/2014940/commnad_line_c2a14.log)
[rainbow_c2a14.log](https://github.com/cmu-mars/issues/files/2014941/rainbow_c2a14.log)

**Comment by ivoysey** [comment]
@pooyanjamshidi @javier-camara @schmerl what's the status here? is this fixed or an RR3 problem?

**Comment by schmerl** [comment]
@javier-camara Pinging you back in on this. It looks like m_scoreboard does not have any entries?

**Comment by schmerl** [comment]
@pooyanjamshidi It looks like the planner couldn't find a plan. We've made some changes so that if that is the case, then we properly report the failure back to you (and the TA). Please rebuild and rerun this test.

**Comment by pooyanjamshidi** [comment]
Logs. 
[rainbow-analyzers.log](https://github.com/cmu-mars/issues/files/2137581/rainbow-analyzers.log)
[rainbow-logging.log](https://github.com/cmu-mars/issues/files/2137582/rainbow-logging.log)

**Comment by schmerl** [comment]
@pooyanjamshidi SOme sort of conclusions would be less obtuse as a comment. I don't see the exception.

**Comment by pooyanjamshidi** [comment]
@schmerl And here is the command line logs. 
[commnad_line-newc1.log](https://github.com/cmu-mars/issues/files/2138236/commnad_line-newc1.log)

Did you see any adaptation in the rainbow log?

**Comment by schmerl** [comment]
@pooyanjamshidi It would be much easier if you told me what you observed happening because  you have the expectation of what you think should happen and also what you observed happening. I think this bug is about the exception. I did not see the exception. If you saw the exception, or you saw something else happening, the be clear about it.

**Comment by pooyanjamshidi** [comment]
@schmerl I didn't see any exception, and the mission was successfully accomplished, but I expected to see some reconfiguration because at the end of the mission the battery was too low and it was risky, makes sense? So in this case, in particular, I expected to see at least a reconfiguration to a conservative configuration, but that didn't happen as far as I can see

**Comment by schmerl** [comment]
@pooyanjamshidi @javier-camara I don't see why you would expect the adaptation to happen if it has enough battery to complete the task, even if you think it was risky. Correct me if I'm wrong (and I don't know what you designed to have happen in this case) but for what you want you would have needed either (a) somehow encoded risk (which I never remember discussing), or (c) encoded some power threshold that you expect to get to a target with (which I also don't remember being in our design discussions). So, from my perspective this particular case works -- the perturbation to reduce the power did not not perturb it significantly enough to warrant Rainbow needing to adapt - it still had enough power to get to the target.

Do you have a case where you've seen Rainbow issue an adaptation because the charge has been perturbed enough?

**Comment by pooyanjamshidi** [comment]
@schmerl Yes, see cases above. 
Probably I don't know the power analyzer much, but does it only issue an adaptation only when for sure the mission cannot be finished with the remaining battery and under current configuration? or it does consider some contingency?

**Comment by schmerl** [comment]
I think if it has enough power to get there, it will think that it is ok
and not need to adapt.

**Comment by bisc** [comment]
I'm out of context here, but it seems like @pooyanjamshidi expected some robust handling of possible noise and model error. Unfortunately, AFAIK, we don't do any such thing now (seems to be a larger engineering burden than we can afford), and we assume the numbers in the prediction/planner to be precise and correct, perfectly aligned with the simulator.

**Comment by pooyanjamshidi** [comment]
@javier-camara In this scenario, the planner issued some adaptation, but the robot went out of battery, can you check why?
[commnad_line-newc2.log](https://github.com/cmu-mars/issues/files/2142258/commnad_line-newc2.log)
[rainbow-analyzers.log](https://github.com/cmu-mars/issues/files/2142259/rainbow-analyzers.log)
[rainbow-logging.log](https://github.com/cmu-mars/issues/files/2142260/rainbow-logging.log)

**Comment by pooyanjamshidi** [comment]
@javier-camara would you please also confirm whether I can adjust the weights for timeliness and energy via the `~/cp1/config.json` as we discussed?
```JSON
{
  "battery_capacity": 32560.0,
  "charging_rate": 69.9555,
  "utility": {
    "energy": 0.5,
    "timeliness": 0.5
  }
}
```

**Comment by pooyanjamshidi** [comment]
In this scenario, the bot ran out of battery. It's strange it plans to charge in the last step.
```P(V(1, do SetCP1Config(1) then 2),V(2, do MoveAbsH(0.54, 10.87, 0.63, 3.1316) then 3)::V(3, do MoveAbsH(-21.08, 11.08, 0.63, -1.5708) then 4)::V(4, do MoveAbsH(-21.08, 0.00, 0.63, -1.5522) then 5)::V(5, do MoveAbsH(-20.86, -11.62, 0.63, -0.0123) then 6)::V(6, do MoveAbsH(1.03, -11.89, 0.63, -0.0123) then 7)::V(7, do Charge(15.00) then 8)::V(8, end)::nil)```



[rainbow-analyzers.log](https://github.com/cmu-mars/issues/files/2143246/rainbow-analyzers.log)
[Uploading rainbow-logging.log…]()
[commnad_line-newc2.log](https://github.com/cmu-mars/issues/files/2143655/commnad_line-newc2.log)

**Comment by schmerl** [comment]
For these bugs can you also upload the console log with all the ros output

**Comment by pooyanjamshidi** [comment]
@schmerl done

**Comment by schmerl** [comment]
@pooyanjamshidi A couple of questions:

1. At the time that the adaptation was issued, what was the charge? I don't see here any traces for the battery, so I can't tell if it is already low when the adaptation occurs.

2. It looks like rainbow-logging.log did not finish uploading before you disconnected, so I can't see what rainbow thinks.

3. @javier-camara What other information would you need to be able to debug this?

**Comment by pooyanjamshidi** [comment]
The battery was around `1.4 wh`. One thought for better debugging: can you guys pull/printout the battery right before the adaptation? this makes debugging better. I don't know about issue number 2.

**Comment by javier-camara** [comment]
Some things with respect to the points discussed in this thread:

- @pooyanjamshidi , I have implemented the customization via configuration file that you mention above. I just pushed the code. Once Bradley goes through the additional step, you should be able to customize things.

- With respect to the bug itself, I have noticed that the time that the IG issues along with the charging instruction is fixed (just 15 seconds). This is the default we had in phase 1. The planner in p2.cp1 knows how much time it's going to take to recharge (it is in the time reward structure and factored into the planning), but we cannot extract that information from the planner. I believe that we discussed this issue in the past. There is some extra code needed, external to the planner, that determines the time required to recharge the full battery, and incorporates this information to the recharging instruction in the IG. This time will only be known once the robot reaches the charging station during execution, and it is (BATTERY_CAPACITY-OBSERVED_BATTERY_CHARGE)/CHARGING_RATE.

**Comment by schmerl** [comment]
So does this mean that the code should be in the instruction to charge or
somewhere in rainbow?

**Comment by pooyanjamshidi** [comment]
@javier-camara do we really need that as in the instruction graph we implemented code that the robot will be fully charged and doesn't need any input. We discussed that before at some point.

**Comment by pooyanjamshidi** [comment]
@javier-camara @schmerl see exception below, the correct path should be: `~/cp1/config.json`.
```cp1_ta    | Configuring rainbow for cp1
cp1_ta    | LD_LIBRARY_PATH=/home/mars/catkin_ws/devel/lib:/opt/ros/kinetic/lib:/opt/ros/kinetic/lib/x86_64-linux-gnu:/home/mars/das/prism-4.3.1-linux64/lib
cp1_ta    | Starting /home/mars/das/rainbow-brass/run-oracle.sh -h -w /home/mars/das/rainbow-brass brass-p2-cp1
cp1_ta    | java.io.FileNotFoundException: /Users/jcamara/Dropbox/Documents/Work/Projects/rainbow-alt/deployments/rainbow-brass/prismtmp/p2cp1/config.json (No such file or directory)
cp1_ta    | 	at java.io.FileInputStream.open0(Native Method)
cp1_ta    | 	at java.io.FileInputStream.open(FileInputStream.java:195)
cp1_ta    | 	at java.io.FileInputStream.<init>(FileInputStream.java:138)
cp1_ta    | 	at java.io.FileInputStream.<init>(FileInputStream.java:93)
cp1_ta    | 	at java.io.FileReader.<init>(FileReader.java:58)
cp1_ta    | 	at org.sa.rainbow.brass.confsynthesis.SimpleConfigurationBatteryModel.loadConfig(SimpleConfigurationBatteryModel.java:59)
cp1_ta    | 	at org.sa.rainbow.brass.confsynthesis.SimpleConfigurationBatteryModel.<init>(SimpleConfigurationBatteryModel.java:84)
cp1_ta    | 	at org.sa.rainbow.brass.plan.p2_cp1.DecisionEngineCP1.init(DecisionEngineCP1.java:29)
cp1_ta    | 	at org.sa.rainbow.brass.adaptation.p2_cp1.CP1BRASSAdaptationPlanner.initialize(CP1BRASSAdaptationPlanner.java:79)
cp1_ta    | 	at org.sa.rainbow.core.RainbowMaster.initializeRainbowComponents(RainbowMaster.java:241)
cp1_ta    | 	at org.sa.rainbow.core.RainbowMaster.initialize(RainbowMaster.java:131)
cp1_ta    | 	at org.sa.rainbow.core.RainbowMaster.main(RainbowMaster.java:798)
cp1_ta    | java.lang.NullPointerException
cp1_ta    | 	at org.sa.rainbow.brass.confsynthesis.SimpleConfigurationBatteryModel.loadConfig(SimpleConfigurationBatteryModel.java:68)
cp1_ta    | 	at org.sa.rainbow.brass.confsynthesis.SimpleConfigurationBatteryModel.<init>(SimpleConfigurationBatteryModel.java:84)
cp1_ta    | 	at org.sa.rainbow.brass.plan.p2_cp1.DecisionEngineCP1.init(DecisionEngineCP1.java:29)
cp1_ta    | 	at org.sa.rainbow.brass.adaptation.p2_cp1.CP1BRASSAdaptationPlanner.initialize(CP1BRASSAdaptationPlanner.java:79)
cp1_ta    | 	at org.sa.rainbow.core.RainbowMaster.initializeRainbowComponents(RainbowMaster.java:241)
cp1_ta    | 	at org.sa.rainbow.core.RainbowMaster.initialize(RainbowMaster.java:131)
cp1_ta    | 	at org.sa.rainbow.core.RainbowMaster.main(RainbowMaster.java:798)
cp1_ta    | Exception in thread "main" org.sa.rainbow.core.error.RainbowConnectionException: Cannot initialize DecisionEngine
cp1_ta    | 	at org.sa.rainbow.brass.adaptation.p2_cp1.CP1BRASSAdaptationPlanner.initialize(CP1BRASSAdaptationPlanner.java:86)
cp1_ta    | 	at org.sa.rainbow.core.RainbowMaster.initializeRainbowComponents(RainbowMaster.java:241)
cp1_ta    | 	at org.sa.rainbow.core.RainbowMaster.initialize(RainbowMaster.java:131)
cp1_ta    | 	at org.sa.rainbow.core.RainbowMaster.main(RainbowMaster.java:798)
cp1_ta    | Caused by: java.lang.NullPointerException
cp1_ta    | 	at org.sa.rainbow.brass.confsynthesis.SimpleConfigurationBatteryModel.loadConfig(SimpleConfigurationBatteryModel.java:68)
cp1_ta    | 	at org.sa.rainbow.brass.confsynthesis.SimpleConfigurationBatteryModel.<init>(SimpleConfigurationBatteryModel.java:84)
cp1_ta    | 	at org.sa.rainbow.brass.plan.p2_cp1.DecisionEngineCP1.init(DecisionEngineCP1.java:29)
cp1_ta    | 	at org.sa.rainbow.brass.adaptation.p2_cp1.CP1BRASSAdaptationPlanner.initialize(CP1BRASSAdaptationPlanner.java:79)
cp1_ta    | 	... 3 more
```

**Comment by pooyanjamshidi** [comment]
@javier-camara @schmerl  in the new version I still see `Charge(15.00)`, as I mentioned, we have the code in IG that simply fully charges, so no need for any specific time being at the charging station. However, in the tests, the robot stays in the charging until it fully charges and `15` doesn't seem to have any effect. Do we need to change the planner or leave it as it is?

**Comment by schmerl** [comment]
If this isn't having an effect, but the robot is charging fully and
behaving as expected, I think we should leave this the way it is. Not only
would it require changes to the planner to change it, it would also require
changes to the instruction graph parser and interpreter that I'd rather not
touch at this late stage.

**Comment by schmerl** [comment]
@pooyanjamshidi Close?

**Comment by pooyanjamshidi** [comment]
@schmerl let's keep this open as I'm running more tests.

**Comment by ivoysey** [comment]
@pooyanjamshidi sounds good. please do close this when you're satisfied, though.

**Comment by pooyanjamshidi** [comment]
With the new update of Rainbow, now the charging plans make more sense, so I'm closing this, I will reopen if found an issue.
